core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java

```diff
@@ -1524,6 +1513,29 @@ private SelfSignedCertificate selfSignedCertificate() throws CertificateExceptio
         return selfSignedCertificate;
     }
 
+    boolean equalsDefaultHostname(String defaultHostname) {
+        return normalizeDefaultHostname(defaultHostname).equals(this.defaultHostname);
+    }
+
+    boolean equalsHostnamePattern(String hostnamePattern) {
+        checkArgument(!hostnamePattern.isEmpty(), "hostnamePattern is empty.");
+
+        final HostAndPort hostAndPort = HostAndPort.fromString(hostnamePattern);
+        if (port == -1 ^ !hostAndPort.hasPort()) {
+            return false;
+        }
+        if (port != -1) {
+            if (port != hostAndPort.getPort()) {
+                return false;
+            }
+            hostnamePattern = hostAndPort.getHost();
+        }
```

Comment:
We may simplify these conditions with `hostAndPort.getPortOrDefault(-1)`.
The code would change if the comment above were addressed though. 
```java
if (port != hostAndPort.getPortOrDefault(-1)) {
    return false;
}
hostnamePattern = hostAndPort.getHost();
```